src/3d/qgs3dsceneexporter.cpp

```diff
@@ -0,0 +1,524 @@
+      indexes.push_back( nextRowStartIndex + i + 1 );
+      indexes.push_back( rowStartIndex + i + 1 );
+    }
+  }
+
+  return indexes;
+}
+
+Qgs3DSceneExporter::Qgs3DSceneExporter( Qt3DCore::QNode *parent )
+  : Qt3DCore::QEntity( parent )
+  , mSmoothEdges( false )
+  , mTerrainResolution( 128 )
+  , mExportNormals( true )
+  , mExportTextures( false )
+  , mTerrainTextureResolution( 512 )
+  , mScale( 1.0f )
+{
+
+}
+
+void Qgs3DSceneExporter::parseEntity( Qt3DCore::QEntity *entity )
+{
+  if ( entity == nullptr ) return;
+  // check if the current entity is a terrain tile
+  QgsTerrainTileEntity *terrainTile = qobject_cast<QgsTerrainTileEntity *>( entity );
+  if ( terrainTile != nullptr )
+  {
+    parseEntity( terrainTile );
+    return;
+  }
+
+  for ( Qt3DCore::QComponent *c : entity->components() )
+  {
+    Qt3DRender::QGeometryRenderer *comp = qobject_cast<Qt3DRender::QGeometryRenderer *>( c );
+    if ( comp == nullptr ) continue;
+    Qt3DRender::QGeometry *geom = comp->geometry();
+
+    QgsTessellatedPolygonGeometry *tessellated = qobject_cast<QgsTessellatedPolygonGeometry *>( geom );
+    if ( tessellated != nullptr )
+    {
+      process( tessellated );
+      continue;
+    }
+
+    //    Qt3DExtras::QPlaneGeometry *plane = qobject_cast<Qt3DExtras::QPlaneGeometry *>( geom );
+    //    if ( plane != nullptr )
+    //    {
+    //      process( plane );
+    //      continue;
+    //    }
+
+  }
+  for ( QObject *child : entity->children() )
+  {
+    Qt3DCore::QEntity *childEntity = qobject_cast<Qt3DCore::QEntity *>( child );
+    if ( childEntity != nullptr ) parseEntity( childEntity );
+  }
+}
+
+void Qgs3DSceneExporter::parseEntity( QgsTerrainEntity *terrain )
+{
+  const Qgs3DMapSettings &settings = terrain->map3D();
+  QgsChunkNode *root = terrain->rootNode();
+
+  // just use LoD0 for now
+  int levelOfDetails = 0;
+  QVector<QgsChunkNode *> leafs;
+  leafs << root;
+  for ( int i = 0; i < levelOfDetails; ++i )
+  {
+    QVector<QgsChunkNode *> nodes = leafs;
+    leafs.clear();
+    for ( QgsChunkNode *node : nodes )
+    {
+      node->ensureAllChildrenExist();
+      QgsChunkNode *const *children = node->children();
+      for ( int i = 0; i < 4; ++i )
+      {
+        if ( children[i] != nullptr )
+        {
+          leafs.push_back( children[i] );
+        }
+      }
+    }
+  }
+
+  QgsTerrainGenerator *generator = settings.terrainGenerator();
+  QgsTerrainTileEntity *terrainTile = nullptr;
+  QgsTerrainTextureGenerator *textureGenerator = terrain->textureGenerator();
+  QSize oldResolution = textureGenerator->textureSize();
+  textureGenerator->setTextureSize( QSize( mTerrainTextureResolution, mTerrainTextureResolution ) );
+  switch ( generator->type() )
+  {
+    case QgsTerrainGenerator::Dem:
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getDemTerrainEntity( terrain, node );
+        this->parseDemTile( terrainTile, textureGenerator );
```

Comment:
Are these tiles getting deleted anywhere?

Reply:
Yes I made the Qgs3DSceneExporter class derive from a QObject so I can make every generated object a child of the scene exporter and it will be deleted automatically after.
It's like I made it into a policy in my mind whenever I make a new object, it should be a child of the object where it is created. That seems like a good idea for our use case.